0x/Contracts/Exchange.cs

```diff
@@ -0,0 +1,151 @@
+﻿
+using Nethereum.Contracts;
+using Nethereum.Hex.HexTypes;
+using Nethereum.RPC.Eth.DTOs;
+using Nethereum.Web3;
+using System;
+using System.Collections.Generic;
+using System.Numerics;
+using System.Threading.Tasks;
+using ZeroX.Enums;
+using ZeroX.Orders;
+using ZeroX.Utilities;
+using Transaction = ZeroX.Transactions.Transaction;
+using Nethereum.Hex.HexConvertors.Extensions;
+
+namespace ZeroX.Contracts
+{
+    public class Exchange : SmartContract
+    {
+        // TODO: Add argument validation, add docs
+
+        static Exchange()
+        {
+            _abi = LoadABI(nameof(Exchange));
+        }
+
+        private static readonly Dictionary<Network, EthereumAddress> _contractAddressses 
+            = new Dictionary<Network, EthereumAddress>
+            {
+                { Network.Main,    (EthereumAddress)"0x4f833a24e1f95d70f028921e27040ca56e09ab0b"},
+                { Network.Ropsten, (EthereumAddress)"0x4530c0483a1633c7a1c97d2c53721caff2caaaaf"},
+                { Network.Rinkeby, (EthereumAddress)"0xbce0b5f6eb618c565c3e5f5cd69652bbc279f44e"},
+                { Network.Kovan,   (EthereumAddress)"0x35dd2932454449b14cee11a94d3674a936d5d7b2"}
+            };
+
+        private static readonly string _abi;
+
+        public Exchange(string rpcUrl, EthereumAddress contractAddress, Account callerAccount)
+            : base(rpcUrl, contractAddress, callerAccount)
+        {
+        }
+
+        public Exchange(string rpcUrl, Network network, Account callerAccount)
+            : this(rpcUrl, _contractAddressses[network], callerAccount)
+        { }
+
+        public async Task<string> FillOrderExecTxAsync(Order order, BigInteger takerAssetFillAmount, 
```

Comment:
Very unspecific and messy (fill order execute transaction). Maybe execute is just the right word. If you try to wrap a smart contract function, just name it the same.

Reply:
You are right, naming is kinda messy, I wanted to differentiate between fillOrder call which is done by taker( hence check for TakerAddress in that function) and fillOrder call made via executeTransaction by third party. I did not want library user to deal with tx data field encoding( hence no  wrapper function for executeTransaction). Mayble All these fillOrder* could be overloaded. Also you could notice there are two types of methods. Those which are static methods, they return Nethereum function and list of parameters (meant to be used with TxProc) and instance methods which call appropriate methods in smart contracts. Nonce counting, gas price estimation must be done manually when using these methods.